test/cpp/util/channelz_sampler_test.cc

```diff
@@ -0,0 +1,167 @@
+    unsigned int server_delay_microseconds = 100000;
+    server_delay_microseconds += rand() % server_delay_microseconds;
+    usleep(server_delay_microseconds);
+    return Status::OK;
+  }
+};
+
+// Run server in a thread
+void RunServer() {
+  // register channelz service
+  ::grpc::channelz::experimental::InitChannelzService();
+
+  EchoServerImpl service;
+  grpc::EnableDefaultHealthCheckService(true);
+  grpc_impl::reflection::InitProtoReflectionServerBuilderPlugin();
+  grpc::ServerBuilder builder;
+  auto server_creds =
+      grpc::testing::GetCredentialsProvider()->GetServerCredentials(
+          custom_credentials_type);
+  builder.AddListeningPort(server_address, server_creds);
+
+  // forces channelz and channel tracing to be enabled.
+  builder.AddChannelArgument(GRPC_ARG_ENABLE_CHANNELZ, 1);
+  builder.AddChannelArgument(GRPC_ARG_MAX_CHANNEL_TRACE_EVENT_MEMORY_PER_NODE,
+                             1024);
+  builder.RegisterService(&service);
+  std::unique_ptr<Server> server(builder.BuildAndStart());
+  std::cout << "Server listening on " << server_address << std::endl;
+
+  server->Wait();
+}
+
+// Creata an echo client - set timeout as 0.15s
+class EchoClientImpl {
+ public:
+  EchoClientImpl(std::shared_ptr<Channel> channel)
+      : stub_(grpc::testing::TestService::NewStub(channel)) {}
+  Status EmptyCall() {
+    grpc::testing::Empty request;
+    grpc::testing::Empty response;
+    ClientContext context;
+    int64_t timeout_microseconds = 150;
+    context.set_deadline(
+        grpc_timeout_milliseconds_to_deadline(timeout_microseconds));
+    Status status = stub_->EmptyCall(&context, request, &response);
+    return status;
+  }
+
+ private:
+  std::unique_ptr<grpc::testing::TestService::Stub> stub_;
+};
+
+// Run client in a thread
+void RunClient(std::string client_id) {
+  // std::string target_str = "localhost:10000";
+  grpc::ChannelArguments channel_args;
+  std::shared_ptr<grpc::ChannelCredentials> channel_creds =
+      grpc::testing::GetCredentialsProvider()->GetChannelCredentials(
+          custom_credentials_type, &channel_args);
+  EchoClientImpl echoer(grpc::CreateChannel(server_address, channel_creds));
+  unsigned int client_echo_sleep_second = 1;
+
+  std::cout << "Client " << client_id << " is echoing!" << std::endl;
+  while (true) {
+    Status status = echoer.EmptyCall();
+    sleep(client_echo_sleep_second);
+  }
+}
+
+// Test the channelz sampler
+TEST(ChannelzSamplerTest, SimpleTest) {
+  // server thread
+  std::thread server_thread(RunServer);
+  std::cout << "Wait 3 seconds to make sure server is started..." << std::endl;
+  float server_start_seconds = 3.0;
+  sleep(server_start_seconds);
+
+  // client thread
+  std::thread client_thread_1(RunClient, "1");
+  std::thread client_thread_2(RunClient, "2");
+  float run_services_seconds = 5.0;
+  std::cout << "Run echo service for " << run_services_seconds << " seconds."
+            << std::endl;
+  sleep(run_services_seconds);
```

Comment:
what's the reason for sleeping here? Can we remove this?

Reply:
Removed